ndpyramid/reproject.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import annotations  # noqa: F401
+
+import typing
+from collections import defaultdict
+
+import datatree as dt
+import numpy as np
+import xarray as xr
+from rasterio.warp import Resampling
+
+from .common import Projection
+from .utils import (
+    add_metadata_and_zarr_encoding,
+    get_levels,
+    get_version,
+    multiscales_template,
+)
+
+
+def _da_reproject(da, *, dim, crs, resampling, transform):
+    if da.encoding.get('_FillValue') is None and np.issubdtype(da.dtype, np.floating):
+        da.encoding['_FillValue'] = np.nan
+    return da.rio.reproject(
+        crs,
+        resampling=resampling,
+        shape=(dim, dim),
+        transform=transform,
+    )
+
+
+def level_reproject(
+    ds: xr.Dataset,
+    *,
+    projection: typing.Literal['web-mercator', 'equidistant-cylindrical'] = 'web-mercator',
+    level: int,
+    pixels_per_tile: int = 128,
+    resampling: str | dict = 'average',
+    extra_dim: str = None,
+    clear_attrs: bool = False,
+) -> xr.Dataset:
+    """Create a level of a multiscale pyramid of a dataset via reprojection.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        The dataset to create a multiscale pyramid of.
+    projection : str, optional
+        The projection to use. Default is 'web-mercator'.
+    level : int
+        The level of the pyramid to create.
+    pixels_per_tile : int, optional
+        Number of pixels per tile
+    resampling : dict
+        Rasterio warp resampling method to use. Keys are variable names and values are warp resampling methods.
+    extra_dim : str, optional
+        The name of the extra dimension to iterate over. Default is None.
+
+    Returns
+    -------
+    xr.Dataset
+        The multiscale pyramid level.
+
+    Warning
+    -------
+    Pyramid generation by level is experimental and subject to change.
+    """
+    save_kwargs = {'pixels_per_tile': pixels_per_tile}
+    attrs = {
+        'multiscales': multiscales_template(
+            datasets=[{'path': '.', 'level': level}],
+            type='reduce',
+            method='pyramid_reproject',
+            version=get_version(),
+            kwargs=save_kwargs,
+        )
+    }
+    dim = 2**level * pixels_per_tile
+    projection_model = Projection(name=projection)
+    dst_transform = projection_model.transform(dim=dim)
+
+    # Convert resampling from string to dictionary if necessary
+    if isinstance(resampling, str):
+        resampling_dict: dict = defaultdict(lambda: resampling)
+    else:
+        resampling_dict = resampling
+
+    # create the data array for each level
+    ds_level = xr.Dataset(attrs=ds.attrs)
+    for k, da in ds.items():
+        if clear_attrs:
+            da.attrs.clear()
+        if len(da.shape) == 4:
```

Comment:
Should this be ≥ 4? or should there be another case if we shave ds.shape == 5?

Reply:
I think it would break for 5+ dim datasets. Perhaps we should add a check and raise an informative error as a separate PR